File: clients/desktop/Core/PairingWire.cs
using System;
using System.Buffers.Binary;
using System.Collections.Generic;
using System.IO;

namespace Mozz.Desktop.Core;

/// <summary>
/// Turns a stream of bytes back into pairing frames, matching
/// <c>Sources/MozzPairing/PairingWire.swift</c>.
/// </summary>
/// <remarks>
/// This is the only part of pairing written twice, and it is deliberately the
/// smallest part. The desktop never decodes a frame's contents — it hands the
/// bytes straight to the core through <c>pairingReceive</c> and gets steps back
/// — so the protocol and the crypto have exactly one implementation. What
/// remains here is a four-byte length prefix, which is small enough to be
/// obviously the same on both sides.
///
/// TCP does not preserve message boundaries, so a frame can arrive in pieces,
/// two frames can arrive in one read, and both can happen at once.
/// </remarks>
public sealed class PairingWire
{
    /// <summary>Matches <c>PairingWire.maxFrameLength</c> in Swift: 8 KiB of
    /// ciphertext plus room for the envelope.</summary>
    public const int MaxFrameLength = 8 * 1024 + 1024;

    private readonly MemoryStream _buffer = new();

    /// <summary>Prefix a payload for the wire.</summary>
    public static byte[] Frame(ReadOnlySpan<byte> payload)
    {
        var framed = new byte[payload.Length + 4];
        BinaryPrimitives.WriteUInt32BigEndian(framed.AsSpan(0, 4), (uint)payload.Length);
        payload.CopyTo(framed.AsSpan(4));
        return framed;
    }

    /// <summary>
    /// Feed whatever arrived; returns every frame that is now complete, which
    /// may be none, one, or several.
    /// </summary>
    public IReadOnlyList<byte[]> Append(ReadOnlySpan<byte> bytes)
    {
        _buffer.Seek(0, SeekOrigin.End);
        _buffer.Write(bytes);

        var held = _buffer.ToArray();
        var frames = new List<byte[]>();
        var offset = 0;

        while (held.Length - offset >= 4)
        {
            // Compare BEFORE narrowing to int. A length with the high bit set
            // casts to a negative int, which sails past a `> MaxFrameLength`
            // check and then indexes out of bounds — so the cast has to come
            // after the limit, not before it. Swift reads this into a 64-bit Int
            // and rejects it correctly, so getting this wrong made the two
            // implementations disagree on exactly the input an attacker picks.
            var claimed = BinaryPrimitives.ReadUInt32BigEndian(held.AsSpan(offset, 4));
            if (claimed > MaxFrameLength)
            {
                // Refuse before reserving anything. A length is a claim, and a
                // claim from a peer we have not authenticated yet is exactly the
                // kind not to act on.
                throw new InvalidDataException($"pairing frame of {claimed} bytes exceeds the limit");
            }
            var length = (int)claimed;
            if (held.Length - offset - 4 < length) break;

            frames.Add(held.AsSpan(offset + 4, length).ToArray());
            offset += 4 + length;
        }

        _buffer.SetLength(0);
        if (offset < held.Length)
        {
            _buffer.Write(held.AsSpan(offset));
        }
        return frames;
    }

    /// <summary>Bytes held back waiting for the rest of their frame.</summary>
    public int PendingByteCount => (int)_buffer.Length;
}

File: clients/android/app/src/main/java/com/thatcube/mozz/pairing/PairingWire.kt
package com.thatcube.mozz.pairing

/**
 * Turns a stream of bytes back into pairing frames, matching
 * `Sources/MozzPairing/PairingWire.swift` and the desktop's `PairingWire.cs`.
 *
 * This is the only part of pairing written on every platform, and it is
 * deliberately the smallest part. Nothing here decodes a frame's contents — the
 * bytes go straight to the core through `pairingReceive` and steps come back —
 * so the protocol and the cryptography have exactly one implementation. What
 * remains is a four-byte length prefix, small enough to be obviously the same
 * on all four sides.
 *
 * TCP does not preserve message boundaries, so a frame can arrive in pieces,
 * two frames can arrive in one read, and both can happen at once.
 */
class PairingWire {

    private var buffer = ByteArray(0)

    /** Bytes held back waiting for the rest of their frame. */
    val pendingByteCount: Int get() = buffer.size

    /**
     * Feed whatever arrived; returns every frame that is now complete, which
     * may be none, one, or several.
     */
    fun append(bytes: ByteArray, count: Int = bytes.size): List<ByteArray> {
        buffer = buffer + bytes.copyOf(count)
        val frames = mutableListOf<ByteArray>()
        var offset = 0

        while (buffer.size - offset >= PREFIX) {
            // Read into a Long and compare BEFORE narrowing. A length with the
            // high bit set becomes a negative Int, which sails past a
            // `> MAX_FRAME_LENGTH` check and then indexes out of bounds. Swift
            // reads this into a 64-bit Int and rejects it correctly, so getting
            // it wrong here would make the two disagree on exactly the input an
            // attacker picks.
            val claimed =
                ((buffer[offset].toLong() and 0xFF) shl 24) or
                    ((buffer[offset + 1].toLong() and 0xFF) shl 16) or
                    ((buffer[offset + 2].toLong() and 0xFF) shl 8) or
                    (buffer[offset + 3].toLong() and 0xFF)
            if (claimed > MAX_FRAME_LENGTH) {
                // Refuse before reserving anything. A length is a claim, and a
                // claim from a peer nobody has authenticated yet is exactly the
                // kind not to act on.
                throw IllegalArgumentException("pairing frame of $claimed bytes exceeds the limit")
            }
            val length = claimed.toInt()
            if (buffer.size - offset - PREFIX < length) break

            frames += buffer.copyOfRange(offset + PREFIX, offset + PREFIX + length)
            offset += PREFIX + length
        }

        buffer = if (offset == 0) buffer else buffer.copyOfRange(offset, buffer.size)
        return frames
    }

    companion object {
        private const val PREFIX = 4

        /** Matches `PairingWire.maxFrameLength` in Swift: 8 KiB plus envelope. */
        const val MAX_FRAME_LENGTH = 8 * 1024 + 1024

        /** Prefix a payload for the wire. */
        fun frame(payload: ByteArray): ByteArray {
            val framed = ByteArray(payload.size + PREFIX)
            framed[0] = (payload.size ushr 24).toByte()
            framed[1] = (payload.size ushr 16).toByte()
            framed[2] = (payload.size ushr 8).toByte()
            framed[3] = payload.size.toByte()
            payload.copyInto(framed, PREFIX)
            return framed
        }
    }
}
